tools/mercury/src/utils/utils.go

```diff
@@ -0,0 +1,306 @@
+    fp ( os.Stderr, "Can't write results file!\n" )
+    os.Exit ( 1 )
+  }
+  defer f.Close ( )
+
+  if test_error == "" {
+    fp ( f, "success\n" )
+  } else {
+    fp ( f, "failure : %s\n", test_error )
+  }
+
+  os.Exit ( 0 )
+}
+
+
+
+
+
+func Make_paths ( mercury_root, test_id, test_name string ) ( router_path, result_path, config_path, log_path string ) {
+  dispatch_install_root := os.Getenv ( "DISPATCH_INSTALL_ROOT" )
+  router_path            = dispatch_install_root + "/sbin/qdrouterd"
+  result_path            = mercury_root + "/results/" + test_name + "/" + test_id
+  config_path            = result_path + "/config"
+  log_path               = result_path + "/log"
+
+  return router_path, result_path, config_path, log_path
+}
+
+
+
+
+
+func Memory_usage ( pid int ) ( rss int ) {
+  proc_file_name := "/proc/" + strconv.Itoa(pid) + "/statm"
+  proc_file, err := os.Open ( proc_file_name )
+  if err != nil {
+    fp ( os.Stderr, "util.Memory_usage error: can't open |%s|\n", proc_file_name )
+    return -1
+  }
+  defer proc_file.Close ( )
+
+  var vm_size int
+  fmt.Fscanf ( proc_file, "%d%d", & vm_size, & rss )
+  return rss
+}
+
+
+
+
+
+func Cpu_usage ( target_pid int ) ( cpu_usage int, err error ) {
+
+  // Let top iterate twice for greater accuracy.
+  command   := "top" 
+  args      := " -b -n 2 -p " + strconv.Itoa ( target_pid )
+  args_list := strings.Fields ( args )
+
+  out, err := exec.Command ( command, args_list... ).Output()
+  if err != nil {
+    return 0, err
+  }
+
+  lines := strings.Split ( string(out), "\n" )
+
+  last_line := lines [ len(lines) - 2 ]
+  fields := strings.Fields ( last_line )
+  cpu_field := fields [ 8 ]
+  temp, err := strconv.ParseFloat ( cpu_field, 32 )
+  if err != nil {
+    return 0, err
+  }
+  cpu_usage = int ( 100 * temp )
+
+  return cpu_usage, nil
+}
+
+
+
+
+
+func Timestamp ( ) ( string ) {
+  now := time.Now()
+  ts1 := now.Format ( "2006-01-02 15:04:05.000000" )
+  nsec := now.UnixNano()
+  fsec := float64(nsec) / 1000000000.0
+  return ts1 + fmt.Sprintf ( " %.6f", fsec )
+}
+
+
+
+
+
+// Use cpupower command to discover allowable frequencies.
+func Get_CPU_freqs ( ) ( freqs [] string ) {
+  command   := "sudo"
+  args      := "cpupower frequency-info"
+  args_list := strings.Fields ( args )
+
+  out, _ := exec.Command ( command, args_list... ).Output()
+
+  lines := strings.Split ( string(out), "\n" )
+  for _, line := range lines {
+    index := strings.Index ( line, "available frequency steps" )
+    if index != -1 {
+      fields := strings.Fields ( line )
+      for _, field := range fields {
+        _, err := strconv.ParseFloat ( field, 32 )
+        if err == nil {
+          freqs = append ( freqs, field ) // I want the string, though. Not the float.
+        }
+      }
+      break
+    }
+  }
+
+  return freqs
+}
+
+
+
+
+
+func Set_Top_Freq ( freqs [] string ) ( err error ) {
+  command := "sudo"
+  args    := "cpupower frequency-set --freq " + freqs [ 0 ] + "GHz"
```

Comment:
This does not pass `go fmt`, /me thinks. The same is true for pretty much all the .go files in the PR.